packages/server/src/orchestrator/state.ts

```diff
@@ -67,7 +69,7 @@ export type ServiceState<O> = {
 } & O;
 
 export type Service<O> = {
-  (options: Partial<Omit<ServiceState<O>, keyof O>> & O): Operation<void>;
+  (options: O & Required<Pick<ServiceState<O>, 'atom'>>): Operation<void>;
 };
```

Comment:
I'm not happy with this type at all. This also feels to me like having something which is halfway done. We're building up patches and bandaids on a type to force it into doing kind of what we want, rather than finding a good generalized solution. If it can't easily be generlized, then maybe it shouldn't be.

Reply:
This comment is a statement.

In order to resolve this, what needs to happen?  There is no indication.  This could be another PR that goes nowhere but I'll wait a few weeks to find that out.

Even better than that, how could we have found this out before arriving at this point after the fact?  

Should the types have been agreed upfront?  

It is a wasted effort on my part to reach this point after.

Reply:
I think @dagda1 is trying to not rush towards an abstraction as that's something that has stalled out PR's in the past. 

I think everyone is 100% not satisfied with service state carrying over artifacts of the app server and proxy server which are very much http service oriented things.

I'm happy to have that discussion now before work proceeds. I think we can model it very simply with

`unstarted`, `starting`, `running`, `stopping`, `stopped` as the discreet states:

```ts
type ServiceState<TOptions, TRunning, TStopped> = 
{
  status: 'unstarted';
} | 
{
  status: 'starting',
  options: TOptions;
} |
{ 
  status: 'running';
  options: TOptions;
  state: TRunning;
} |
{ 
  status: 'stopping';
  options: TOptions;
  state: TRunning;
} | 
{
  status: 'stopped';
  toptions: TOptions;
  state: TRunning;
  stop: TStopped;
}
```

The nice thing about this is that error and crash exits can be modeled with a custom `TStop` for example:

```ts
type AppServiceState = ServiceState<AppOptions, AppState, ExitStatus>
```

Reply:
I've absolutely no problem with this being queried or objected to, I don't really like the union TBH but was happy to just see how it played routine in subsequent PRs under the guise of destroying the Mailboxes.

Getting it right the first time is difficult without fully understanding the services.

I think my point is that this is potentially a showstopper six days after raising the PR with no clear way to resolve it and then the PR just drifts due to our truly remote nature across timezones.

I've always felt git diffs are poor code reviews, I find it hard to get the big picture and only really notice small things.

I totally actually agree with the type being not up to scratch but I was going with the premise that it is good enough for now.

Reply:
I can tell you're frustrated, but let's not call the PR adrift until it's actually adrift. It's true that it sat for most of last week while we were really laboring to close out remaining bugs for our beta milestone, but the moment that there was bandwidth this week, it got picked up and review started.

Still, it does suck to feel like the goal-posts have gotten moved to an undisclosed location. So how about this, what if we just didn't generalize over service states at all and say that the first step is to just get the state in the atom in the first place and use a one-off API for each service. I think that would be a clear win.

Reply:
definitely not frustrated with the points raised.  All good points

Just that it took a while for it to surface

I think in future, I would create a typescript playground of the types and put it in discord before progressing.

Reply:
@dagda1 I'm sorry, reading this again you're totally right, my comment was not very actionable, and doesn't help in closing out this PR. I should have been clearer in my feedback.

I don't see this as stalled at all though, barring a few minor changes, I think this is pretty close. I do like this PR a lot, I'm very excited about this new service architecture, it's a very cool use of Effection, and I think it's going to be way nicer than what we have now.

As for getting timely feedback I don't work on the weekends, so the initial review I did was as timely as I could have given it, the feedback that I posted yesterday was after looking at this more thoroughly, and spotting some issues I hadn't seen before.

The heart of my comment here is that we started off with all of the service interfaces being pretty specific to the app server, which is fine, since it was the only service that we had. But once we move towards adding more services, IMO we should try finding an abstraction which works for all of them.

Reading my comment above it reads much more broadly than what I meant. I was specifically talking about the state of a service, where we've tried to create an abstraction trying to generalize what the state of a service looks like, but maybe this is not easily generalized, because the various services that we have have pretty different needs as to the state that they should maintain.

I've slightly modified the playground link you posted in Discord to illustrate what I mean:

https://codesandbox.io/s/typescript-playground-export-forked-zgx3n?file=/index.ts

In summary:

``` typescript
type AppStatus = { status: "pending" } | { status: "ready" } | { status: "exited"; exitStatus: ExitStatus };
type AppOptions = { url?: string; command?: string; };

interface ServiceStatus {
  status: string;
}

type ServiceState<O, S extends ServiceStatus> = {
  id: string;
  status: S;
  options: O;
};

type Service<O, S extends ServiceStatus> = {
  (status: Slice<S>, options: O & { atom: Slice<OrchestratorState> }): void;
};

type OrchestratorState = {
  appService: ServiceState<AppOptions, AppStatus>;
};

const appService: Service<AppOptions, AppStatus> = (status, options) => {
  status.set({ status: "starting" });
  console.log(options.atom.get().other);
};
```

The key difference here is that we don't try to create an abstraction over the status of various services, but allow them to specify their own status options. I think this makes this whole thing much easier to reason about, and makes it easy to see a path forward for this.

I think it could be sensible to add some constraint to the `status` in the future, maybe `pending` and `ready` or whatever we want to call them, should be defined for all services. Or maybe just `ready`? But it feels like we can probably hold off on this constraint until we know more about how we want to use services. I think just going with convention here is fine for now.

I also made some other minor modification which I think are nice, but these are just suggestions:

- Passing in the status slice directly to the Service, since modifying this will be a key part of the service.
- Removing the atom from the state (since that would be circular), and instead only adding it to the options passed to the service. The exact interface here is just a suggestion.
- Keeping the `options` separate, instead of in the same object makes it easier to extract them and pass them to the service.

Reply:
@jnicklas no need to apologise.  My reply could definitely have been worded better as I wrote with an air of frustration :).

At no point did I agree with the points raised.   I'm in agreement with pretty much all of them

> Passing in the status slice directly to the Service, since modifying this will be a key part of the service.

@cowboyd and I did think about this but then thought about the case where you might need more than one slice.  I think I should have waited and crossed this bridge when I encountered it.

> Removing the atom from the state (since that would be circular), and instead only adding it to the options passed to the service. The exact interface here is just a suggestion.

This is actually the way it was but my type used `Omit` and other shenanigans so that it was passed into the service but not in the state.

I think the type read terribly in retrospect so I need to make that clearer.

> Keeping the options separate, instead of in the same object makes it easier to extract them and pass them to the service.

I think this makes sense.

Reply:
> Or maybe just ready

I was thinking along these lines also.  We might just need:

```ts
type ServiceStatus = 'IDLE' | 'READY'; 
```

This is good enough for now. 

cc/ @cowboyd

Reply:
@jnicklas @cowboyd I agree with the suggestion that each service having its own status is a better idea.

```ts
export type ServiceStatus = {
  type: string;
};

export type ServiceState<S extends ServiceStatus, O> = {
  options: O;
  status: S;
};

export type Service<S extends ServiceStatus, O> = {
  (status: Slice<S, OrchestratorState>, options: O & { atom: Atom<OrchestratorState> }): Operation<void>;
};
```